app/src/main/java/org/cru/godtools/ui/dashboard/tools/ToolsFragment.kt

```diff
@@ -2,24 +2,116 @@ package org.cru.godtools.ui.dashboard.tools
+import org.cru.godtools.ui.tools.ToolsAdapter
+import org.cru.godtools.ui.tools.ToolsAdapterCallbacks
+import org.cru.godtools.ui.tools.ToolsAdapterViewModel
 
 @AndroidEntryPoint
-class ToolsFragment : BasePlatformFragment<DashboardToolsFragmentBinding>(R.layout.dashboard_tools_fragment) {
+class ToolsFragment :
+    BasePlatformFragment<DashboardToolsFragmentBinding>(R.layout.dashboard_tools_fragment),
+    ToolCategoriesAdapter.Callbacks,
+    ToolsAdapterCallbacks {
+    @Inject
+    internal lateinit var downloadManager: GodToolsDownloadManager
 
-    private val dataModel: ToolsFragmentDataModel by viewModels()
-    override fun onBindingCreated(binding: DashboardToolsFragmentBinding, savedInstanceState: Bundle?) {
-        super.onBindingCreated(binding, savedInstanceState)
-        binding.refresh.setupSwipeRefresh()
-        binding.hasSpotlight = dataModel.hasSpotlight
+    // region Data Model
+    private val dataModel: ToolsCategoryDataModel by viewModels()
+    private val toolsDataModel: ToolsAdapterViewModel by viewModels()
+    // endregion Data Model
+
+    private val categoryAdapter: ToolCategoriesAdapter by lazy {
+        ToolCategoriesAdapter(
+            this,
+            dataModel.selectedCategory,
+            dataModel.primaryLanguage
+        ).also { adapter ->
+            adapter.callbacks.set(this)
+            lifecycle.onDestroy { adapter.callbacks.set(null) }
+            dataModel.categories.observe(this, adapter)
+        }
+    }
+
+    private val toolsAdapter: ToolsAdapter by lazy {
+        ToolsAdapter(this, toolsDataModel, R.layout.dashboard_list_item_tool).also { adapter ->
+            adapter.callbacks.set(this)
+            lifecycle.onDestroy { adapter.callbacks.set(null) }
+            dataModel.filteredTools.observe(this, adapter)
+        }
+    }
+
+    private val toolsSpotlightAdapter: ToolsAdapter by lazy {
+        ToolsAdapter(this, toolsDataModel, R.layout.dashboard_tools_spotlight_list_item).also { adapter ->
+            adapter.callbacks.set(this)
+            lifecycle.onDestroy { adapter.callbacks.set(null) }
+            dataModel.spotlightTools.observe(this, adapter)
+        }
+    }
+
+    private fun createCombinedAdapter() = ConcatAdapter().also { concatAdapter ->
+        concatAdapter.addLayout(R.layout.dashboard_spotlight_concat) {
+            it.findViewById<RecyclerView>(R.id.concatRecyclerView).adapter = toolsSpotlightAdapter
+        }.also { spotlight ->
+            dataModel.hasSpotlight.observe(this) { spotlight.repeat = if (it) 1 else 0 }
+        }
+        concatAdapter.addLayout(R.layout.dashboard_tools_ui_categories) {
+            it.findViewById<RecyclerView>(R.id.concatRecyclerView).adapter = categoryAdapter
+        }.also { category ->
+            dataModel.hasCategories.observe(this) {
+                category.repeat = if (it) 1 else 0
+            }
+        }
+        concatAdapter.addAdapter(toolsAdapter)
     }
```

Comment:
this doesn't solve the potential memory leak issue, the leak would come from the toolsAdapter and toolsSpotlightAdapter. those would maintain a reference to the `RecyclerView` in the inflated views.